rfcs/rdm-0079-collections.md

```diff
@@ -0,0 +1,337 @@
+            _path = ","
+        
+        # INSERT INTO collection 
+        ...
+
+    @classmethod
+    def get(cls, id_):
+        """Primary key lookup."""
+        ...
+    
+    @cached_property
+    def sub_collections(self):
+        """Fetch all the descendants."""
+        return self.get_subcollections()
+    
+    @cached_property
+    def direct_subcollections(self):
+        """Fetch only direct descendants."""
+        return self.get_direct_subcollections()
+    
+    def get_direct_subcollections(self):
+        """More preformant descendants query, by path matching."""
+        '''
+            SELECT * 
+            FROM collections
+            WHERE path = self.path + self.id + ","
+            ORDER BY (path, order)
+        '''
+        ...
+    
+    def get_subcollecions(self, max_depth=2):
+        """Get all the subcollections using LIKE query, with the path anchored to the left."""
+        '''
+            SELECT * 
+            FROM collections
+            WHERE path like 'self.path + self.id + ',%'
+            ORDER BY (path, order)
+        '''
+        ...
+      
+    
+    @cached_property
+    def ancestors(self):
+        """
+            Split path by comma (',') and lookup each part of the path individually.
+        """
+        if not self.model:
+            return None
+
+        cps = self.path.split(",")
+        ret = []
+        for cid in cps:
+            if not cid:
+                continue
+            cl = Collection.get(cid)
+            cl.community = self._community
+            ret.append(cl)
+        return list(sorted(ret, key=lambda x: (x.path, x.order)))
+```
+
+The main goals of the programmatic API are:
+    
+1. **Abstraction** - abstract the end user of terms such as "path", how the path is computed, etc, and let the end user focusing on what really matters: work with collections (e.g. render them on the UI)
+2. **Encapsulation** - restrict the user of some operations, e.g. DB queries, so they can be performed in a consistent way efficiently
+
+## Alternatives
+    
+**Pattern reference**
+    
+We have also considered using other patterns, e.g.  [pattern reference](https://www.mongodb.com/docs/manual/tutorial/model-tree-structures-with-parent-references/). 
+    
+In our case, pattern references could work, but the queries would be harder to understand:
+    
+- for example, fetching all descendants of a node (collection) would require more complex logic than a simple text match query. 
+- although recursive queries can retrieve all the ancestors of a node with similar performance to a materialized path, they are more difficult to comprehend and implement.
+
+## Unresolved questions
+    
+1. **Non-community collections**: the endpoint is not yet defined.
+
+2. **Support alternative query strategies**: While we have defined query concatenation as the default approach, there may be other use cases where the query should be composed as a union of all descendant queries.​
+
```

Comment:
this document is not touching the notion of collections built on facets, which we have discussed, is it on purpose?

Reply:
Facets per collection needs more work and iterations, because storing facets in the DB can lead to weird errors if e.g. a facet is updated. There are other questions, e.g. `where does the collection facet come from? ` - it could be from itself or from parents in the tree.

You are right, it should be mentioned in the RFC and added to the `Open questions`.